lib/parser.ts

```diff
@@ -0,0 +1,414 @@
+interface ParsedElement {
+  pcb_id?: string
+  parser?: Parser
+  resolution?: Resolution
+  unit?: Unit
+  structure?: any
+}
+
+interface Parser {
+  space_in_quoted_tokens?: string
+  host_cad?: string
+  host_version?: string
+  constant?: {
+    [key: string]: string
+  }
+}
+
+interface Resolution {
+  unit: string
+  value: string
+}
+
+type Unit = string
+
+function parseDsnElement(dsn: any[]): ParsedElement {
+  const parsedElement: ParsedElement = {
+    pcb_id: dsn[1],
+  }
+
+  dsn.slice(1).forEach((element) => {
+    const elementType = element[0]
+    switch (elementType) {
+      case "pcb":
+        parsedElement.pcb_id = element[1]
+        break
+      case "parser":
+        parsedElement.parser = parseParser(element.slice(1))
+        break
+      case "resolution":
+        parsedElement.resolution = {
+          unit: element[1],
+          value: element[2],
+        }
+        break
+      case "unit":
+        parsedElement.unit = element[1]
+        break
+      case "structure":
+        parsedElement.structure = parseStructure(element.slice(1))
+        break
+      // Add more cases for other element types if needed
+    }
+  })
+
+  return parsedElement
+}
+
+function parseParser(element: any[]): Parser {
+  const parserElement: Parser = {}
+
+  element.forEach((e) => {
+    if (e[0] === "space_in_quoted_tokens") {
+      parserElement.space_in_quoted_tokens = e[1]
+    } else if (e[0] === "host_cad") {
+      parserElement.host_cad = e[1]
+    } else if (e[0] === "host_version") {
+      parserElement.host_version = e[1]
+    } else if (e[0] === "constant" && Array.isArray(e[1])) {
+      // Initialize the constant object if it's not already initialized
+      if (!parserElement.constant) {
+        parserElement.constant = {}
+      }
+      // Add the constant key-value pair
+      parserElement.constant[e[1][0]] = e[1][1]
+    }
+  })
+
+  return parserElement
+}
+
+function parseStructure(elements: any[]): any {
```

Comment:
can we get better types here? all the names are so generic, it's really easy to lose track of what's going on